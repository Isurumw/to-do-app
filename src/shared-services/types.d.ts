declare interface IStorage {
  get(key: string): Promise<string | undefined>;
  getObject<P>(key: string): Promise<P | undefined>;
  set(key: string, value: string): Promise<void>;
  remove(key: string): void;
}

declare interface INetworkManager {
  request(
    route: string,
    method: 'POST' | 'GET' | 'PUT' | 'DELETE' | 'PATCH',
    variables?: object | null,
  ): any;
}
