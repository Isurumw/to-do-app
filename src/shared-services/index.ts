import Config from 'react-native-config';
import AsyncStorage from '@react-native-community/async-storage';

import _NetworkManager from './network-services/network-manager';
import _StorageService from './storage-service/storage-service';

const StorageService = new _StorageService(AsyncStorage);
const NetworkManager = new _NetworkManager(
  Config.AUTH_BASE,
  Config.AUTH_KEY,
  StorageService,
  fetch,
);
const FirestoreManager = new _NetworkManager(
  Config.FIRESTORE_BASE,
  Config.AUTH_KEY,
  StorageService,
  fetch,
);

export {NetworkManager, FirestoreManager, StorageService};
