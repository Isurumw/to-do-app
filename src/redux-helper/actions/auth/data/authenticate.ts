import {NetworkManager, StorageService} from 'shared-services';
import {AUTH_TOKEN_KEY} from 'utils/constants';

type Response = {
  kind: string;
  localId: string;
  email: string;
  idToken: string;
  registered: boolean;
};

const authenticate = async (email: string, password: string) => {
  const response = await NetworkManager.request<Response>(
    'accounts:signInWithPassword',
    'POST',
    {email: email, password: password},
  );

  const {idToken, registered} = response;

  await StorageService.set(AUTH_TOKEN_KEY, idToken);

  return registered;
};

export default authenticate;
