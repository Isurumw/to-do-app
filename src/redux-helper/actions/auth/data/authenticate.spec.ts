import {NetworkManager, StorageService} from 'shared-services';
import authenticate from './authenticate';

jest.mock('shared-services/network-services/network-manager');
const mockedNetworkManager =
  NetworkManager as unknown as jest.Mocked<INetworkManager>;

jest.mock('shared-services/storage-service/storage-service');
const mockedStorageService = StorageService as unknown as jest.Mocked<IStorage>;

describe('check the behaviour of the authenticate natwork service', () => {
  afterEach(() => {
    mockedNetworkManager.request.mockClear();
    mockedStorageService.set.mockClear();
  });

  it('user authentication successful', async () => {
    mockedStorageService.set = jest.fn();

    mockedNetworkManager.request.mockImplementation(() => ({
      kind: 'identitytoolkit#VerifyPasswordResponse',
      localId: 'gbl9goSb0cWtVPeS7HA7was54Q02',
      email: 'isuru@gmail.com',
      displayName: '',
      idToken: 'mock_token',
      registered: true,
    }));

    const response = await authenticate('mock_email', 'mock_password');

    expect(mockedNetworkManager.request).toBeCalledTimes(1);
    expect(mockedStorageService.set).toBeCalledTimes(1);
    expect(response).toEqual(true);
  });

  it('user authentication failed with an error', async () => {
    mockedNetworkManager.request.mockImplementation(() => {
      throw new Error('Some API Error');
    });

    expect(
      authenticate('mock_email', 'mock_password'),
    ).rejects.toThrowErrorMatchingSnapshot();
  });
});
