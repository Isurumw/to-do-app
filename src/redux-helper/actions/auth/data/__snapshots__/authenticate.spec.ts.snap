// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check the behaviour of the authenticate natwork service user authentication failed with an error 1`] = `"Some API Error"`;
